src/traces/heatmap/clean_2d_array.js

```diff
@@ -30,12 +31,38 @@ module.exports = function clean2dArray(zOld, transpose) {
         old2new = function(zOld, i, j) { return zOld[i][j]; };
     }
 
+    var padOld2new = function(zOld, i, j) {
+        if(i === undefined || j === undefined) return undefined;
+        return old2new(zOld, i, j);
+    };
+
+    function axisMapping(ax) {
+        if(trace && trace.type !== 'carpet' && trace.type !== 'contourcarpet' &&
+            ax && ax.type === 'category' && trace['_' + ax._id.charAt(0)].length) {
+            var axMapping = [];
+            for(i = 0; i < ax._categories.length; i++) {
+                axMapping.push((trace['_' + ax._id.charAt(0) + 'Map'] || trace[ax._id.charAt(0)]).indexOf(ax._categories[i]));
```

Comment:
adding `var axLetter = ax._id.charAt(0)` above probably wouldn't hurt.

Reply:
Oh and can we use `ax._categoriesMap` instead of a (potentially) costly `indexOf` into `ax._categories`?

Reply:
> Oh and can we use `ax._categoriesMap` instead of a (potentially) costly `indexOf` into `ax._categories`?

Unfortunately, we cannot. For example, in the `heatmap_shared_categories` mock, `_categoriesMap` links category `Team C` with index 2. That index is out of range for the second trace since it only has 2 columns. We really need to check whether each trace has data for a given category (and fill with `BADNUM`) to fix #1117.

Reply:
> adding `var axLetter = ax._id.charAt(0)` above probably wouldn't hurt.

Done in ebf545ff33bd2f9fee38b516d73c891a2a2d1bee

Reply:
> Ok, that's very unfortunate, because
> 
> ```js
> for(i = 0; i < ax._categories.length; i++) {
>   trace[ax._id.charAt(0)]).indexOf(ax._categories[i]))
> }
> ```
> 
> can be very costly.

What we could do instead is build a `_traceCategoriesMap` for each trace. It would map categories to their position within the trace-level `z` data. This might be a better idea now that I think about it.

Reply:
@etpinard I think can remove `indexOf` with the following:
```javascript
for(i = 0; i < trace[axLetter].length; i++) {
    axMapping[trace[axLetter][i]] = i;
}
return function(i) {return axMapping.hasOwnProperty(ax._categories[i]) ? axMapping[ax._categories[i]] : BADNUM;};
```
Would this be good enough?

Reply:
I went ahead and committed the suggested change above in f180c79 !

Reply:
I didn't know you could use `hasOwnProperty` on an array. Two concerns:

- Is that support across browsers? 
- is that really faster than `indexOf`?

Reply:
What if instead you made `_(x|y)CategoryMap` lookup objects? 

That is, instead of:

```js
   if(ax1 && ax1.type === 'category') {
        trace['_' + var1Name + 'CategoryMap'] = col1vals.map(function(v) { return ax1._categories[v];});
    }
```

we could have

```js
   if(ax1 && ax1.type === 'category') {
        var lookup = trace['_' + var1Name + 'CategoryMap'] = {};
        for(var i = 0; i < col1vals.length; i++) {
           var v = col1vals[i];
           lookup[v] = i;
       }
    }
```

Reply:
About your questions regarding `hasOwnProperty`:
- we have it in several places already and it seems [well supported](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty#Browser_compatibility)
- `hasOwnProperty` is roughly 6 times faster than `indexOf`

However, we now run it for every element so in the end, I'm not sure which is faster.

Reply:
Sorry @antoinerg I got confused (again). Your logic looks right to me. 

The way you have it setup right now, I _think_ we can get rid of the `axMapping.hasOwnProperty(ax._categories[i])` and just have:

```js
function(i) { 
  var ind = axMapping[ax._categories[i]];
  return ind + 1 ? ind : BADNUM;
};
```

where the `+ 1` takes care of the `ind === 0` case. 

This comment is non-blocking as I don't expect much of a performance gain from it :smirk:

Reply:
> where the `+ 1` takes care of the `ind === 0` case.

Awesome trick :) Done in c7cd1f3

Reply:
:dancer: :dancer: